src/middleware/subscribeWithSelector.ts

```diff
@@ -1,43 +1,62 @@
 import {
-  EqualityChecker,
-  GetState,
-  SetState,
   State,
+  StateCreator,
   StateListener,
-  StateSelector,
-  StateSliceListener,
   StoreApi,
+  StoreMutatorIdentifier,
   Subscribe,
 } from '../vanilla'
 
-export type StoreApiWithSubscribeWithSelector<T extends State> = StoreApi<T> & {
-  subscribe: Subscribe<T> & {
-    (listener: StateListener<T>): () => void
-    <StateSlice>(
-      selector: StateSelector<T, StateSlice>,
-      listener: StateSliceListener<StateSlice>,
-      options?: {
-        equalityFn?: EqualityChecker<StateSlice>
-        fireImmediately?: boolean
-      }
-    ): () => void
+type SubscribeWithSelector = <
+  T extends State,
+  Mps extends [StoreMutatorIdentifier, unknown][] = [],
+  Mcs extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  initializer: StateCreator<
+    T,
+    [...Mps, ['zustand/subscribeWithSelector', never]],
+    Mcs
+  >
+) => StateCreator<T, Mps, [['zustand/subscribeWithSelector', never], ...Mcs]>
+
+declare module '../vanilla' {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface StoreMutators<S, A> {
+    ['zustand/subscribeWithSelector']: WithSelectorSubscribe<S>
   }
 }
 
-export const subscribeWithSelector =
-  <
-    S extends State,
-    CustomSetState extends SetState<S>,
-    CustomGetState extends GetState<S>,
-    CustomStoreApi extends StoreApi<S>
-  >(
-    fn: (set: CustomSetState, get: CustomGetState, api: CustomStoreApi) => S
-  ) =>
-  (
-    set: CustomSetState,
-    get: CustomGetState,
-    api: CustomStoreApi & StoreApiWithSubscribeWithSelector<S>
-  ): S => {
+type WithSelectorSubscribe<S> = S extends { getState: () => infer T }
+  ? S & StoreSubscribeWithSelector<Extract<T, State>>
+  : never
+
+interface StoreSubscribeWithSelector<T extends State> {
```

Comment:
Can this be type alias?

Reply:
It can be but I prefer interfaces wherever possible for better quickinfo tooltips and error messages.

Reply:
I prefer type aliases when we don't intend to overload to avoid accidental overloading (but I know this is not very popular opinion and controversial.)
I didn't know about the difference in tooltips/messages.

Reply:
Not here but usually there is a considerable amount of performance benefit too, as with interface typescript knows it will strictly be an object, with type alias it can be anything—unions, intersections, conditions, etc.

Should we go with interfaces? The biggest benefit is with error messages.

Reply:
I wasn't aware about the performance benefit with being an object.

Well, I may change my ming in the future, but let's keep it an interface for this PR.
(I need to experiment the error message difference to better understand.)